src/orchard/xref.clj

```diff
@@ -13,21 +32,87 @@
-    (set (some->> val class .getDeclaredFields
+(defn- f->sym [^clojure.lang.AFn f]
+  (-> f .getClass .getName repl/demunge symbol))
+
+(defn- fn-source [f]
+  (hunt-down-source (f->sym f)))
+
+(defn- fn-name [^java.lang.Class f]
+  (-> f .getName repl/demunge symbol))
+
+(defonce classbytes (atom {}))
+
+(defn- recompile [ns-sym form]
+  (push-thread-bindings
+   {clojure.lang.Compiler/LOADER
+    (proxy [clojure.lang.DynamicClassLoader] [@clojure.lang.Compiler/LOADER]
+      (defineClass
+        ([name bytes src]
+         (swap! classbytes assoc name bytes)
+         (proxy-super defineClass name bytes src))))})
+  (try
+    (let [line @clojure.lang.Compiler/LINE
+          column @clojure.lang.Compiler/COLUMN
+          line (if-let [line (:line (meta form))]
+                 line
+                 line)
+          column (if-let [column (:column (meta form))]
+                   column
+                   column)]
+      (push-thread-bindings {clojure.lang.Compiler/LINE line
+                             clojure.lang.Compiler/COLUMN column})
+      (try
+        (let [form (macroexpand form)]
+          (when (and (coll? form) (= 'clojure.core/fn (first (nth form 2 nil))))
+            (binding [*ns* (create-ns ns-sym)]
+              (clojure.lang.Compiler/analyze
+               clojure.lang.Compiler$C/EVAL
+               (nth form 2)))))
+        (finally
+          (pop-thread-bindings))))
+    (finally
+      (pop-thread-bindings))))
+
+(defn- fn-deps-class
+  [v]
+  (let [^java.lang.Class v (if (class? v)
+                             v
+                             (eval v))]
+    (set (some->> v .getDeclaredFields
                   (keep (fn [^java.lang.reflect.Field f]
                           (or (and (identical? clojure.lang.Var (.getType f))
                                    (java.lang.reflect.Modifier/isPublic (.getModifiers f))
                                    (java.lang.reflect.Modifier/isStatic (.getModifiers f))
                                    (-> f .getName (.startsWith "const__"))
-                                   (.get f val))
+                                   (.get f (fn-name v)))
                               nil)))))))
 
+(defn fn-deps-compiling
+  "Returns a set with all the functions invoked by `val`.
+  `val` can be a function value, a var or a symbol."
+  {:added "0.5"}
+  [f]
+  (reset! classbytes {})
+  (let [f (as-val f)]
+    (when-let [source (and (fn? f) (fn-source f))]
+      (let [sym (f->sym f)
+            b (do (recompile (-> sym namespace symbol) source) @classbytes)
+            class-names (map first b)
+            deps (set (mapcat #(-> % symbol fn-deps-class) class-names))]
+        deps))))
+
+(defn fn-deps [s]
+  (when-let [^clojure.lang.AFn v (as-val s)]
+    (let [f-class-name (-> v .getClass .getName)
+          field (->> clojure.lang.DynamicClassLoader .getDeclaredFields second)
+          classes (into {} (.get field clojure.lang.DynamicClassLoader))
```

Comment:
this holds all functions compiled by clojure:
{...
"dev$dummy"
   #object
    [java.lang.ref.SoftReference 0x5081395f
     "java.lang.ref.SoftReference@5081395f"]
...
}
where lambdas are named "dev$dummy$fn_1234"

the filter uses that fact to just keep the function + lambda classes.

The problem is, when you recompile a function, new lambda classes get generated and the old ones stay around.

Reply:
>  the old ones stay around

As in, garbage to be collected?

What causes that retention? This is a `defn` (and not e.g. a `def`), so why invoking a defn would retain memory references?

Unless there's global state somewhere, probably your solution is good - it's just minor details like `*1` or other repl state that might retain objects for longer than expected.

Reply:
I probably explained it wrong.
The problem is not the memory, the problem is that the Clojure classloader keeps the old classes in its own cache as java.lang.ref.SoftReference and only deletes that if you try to access the reference after it has been cleaned up.
So if I do the following in my code:
```clojure
(defn dummy []
  (map #(println %) (range 10)))
(defn dummy []
  (map #(print %) (range 10)))
;;o r just send both versions to the repl by changing the first
```
classes will now contain "dev$dummy$fn_123" and "dev$dummy$fn_124" where 123 references println and 124 references print.
I haven't found an easy way to get rid of "dev$dummy$fn_123" yet.

Reply:
Perhaps it would be solved if WeakReference was used instead.

However, as mentioned personally I do not condone classloader hacking :)

Reply:
This solution does not overwrite the clojure classLoader, so I can't switch the reference type. This just accesses the regular clojure loader through reflection.
I probably shouldn't have kept two separate solutions in the same file, I can see how that is confusing, sorry.

Reply:
> I haven't found an easy way to get rid of "dev$dummy$fn_123" yet.

Does the PR still have this problem?

It might be acceptable to have that problem, if so at least leave a comment explaining what's going on

Reply:
Yes, the PR still has the problem. I'll play around with it some more to see if I can find a solution. If I can't find one I'll add a comment describing the issue and open a new ticket once the PR is merged.